integration/test/Test/Conversation.hs

```diff
@@ -1028,3 +1028,81 @@ testGetSelfMember = do
     resp.json %. "service" `shouldMatch` Null
     resp.json %. "status" `shouldMatchInt` 0
     resp.json %. "status_ref" `shouldMatch` "0.0"
+
+testReplaceMembers :: (HasCallStack) => App ()
+testReplaceMembers = do
+  [alice, bob, charlie, dylan] <- createAndConnectUsers [OwnDomain, OwnDomain, OwnDomain, OwnDomain]
+  conv <- postConversation alice (defProteus {qualifiedUsers = [bob, charlie]}) >>= getJSON 201
+
+  -- Replace members: remove bob, add dylan, keep charlie
+  [charlieId, dylanId] <- for [charlie, dylan] (%. "qualified_id")
+  bindResponse (replaceMembers alice conv def {users = [charlieId, dylanId]}) $ \resp -> do
+    resp.status `shouldMatchInt` 200
+    resp.json %. "type" `shouldMatch` "conversation.member-join"
+    resp.json %. "qualified_from" `shouldMatch` objQidObject alice
+    resp.json %. "qualified_conversation" `shouldMatch` objQidObject conv
+    users <- resp.json %. "data.users" >>= asList
+    addedUsers <- forM users (%. "qualified_id")
+    addedUsers `shouldMatchSet` [dylanId]
+
+  -- Verify conversation members
+  bindResponse (getConversation alice conv) $ \resp -> do
+    resp.status `shouldMatchInt` 200
+    mems <- resp.json %. "members.others" & asList
+    memIds <- forM mems (%. "qualified_id")
+    memIds `shouldMatchSet` [charlieId, dylanId]
+
+testReplaceMembersUnchanged :: (HasCallStack) => App ()
+testReplaceMembersUnchanged = do
+  [alice, bob, charlie] <- createAndConnectUsers [OwnDomain, OwnDomain, OwnDomain]
+  conv <- postConversation alice (defProteus {qualifiedUsers = [bob, charlie]}) >>= getJSON 201
+
+  -- Replace members with the same set
+  [bobId, charlieId] <- for [bob, charlie] (%. "qualified_id")
+  bindResponse (replaceMembers alice conv def {users = [bobId, charlieId]}) $ \resp -> do
+    resp.status `shouldMatchInt` 204
```

Comment:
also look at the returned member set here?

Reply:
added, thanks!